bft_interp/src/lib.rs

```diff
@@ -1,39 +1,510 @@
             tape_can_grow,
+            program,
+            program_counter: 0,
+        }
+    }
+
+    /// Interpret the program
+    ///
+    /// ```no_run
+    ///# fn main() -> Result<(), Box<dyn std::error::Error>>{
+    ///# use bft_types::BfProgram;
+    ///# use bft_interp::VirtualMachine;
+    ///# use std::num::NonZeroUsize;
+    ///#
+    /// let bf_program = BfProgram::from_file("my_bf_program.bf")?;
+    ///
+    /// let tape_size: Option::<NonZeroUsize> = Some(NonZeroUsize::new(30000).unwrap());
+    /// let mut bf_interpreter: VirtualMachine<u8> = VirtualMachine::new(&bf_program, tape_size, true);
+    /// bf_interpreter.interpret_program()?;
+    ///#
+    ///# Ok(())
+    ///# }
+    /// ```   
+    pub fn interpret_program(&mut self) -> Result<(), VMError> {
+        while self.program_counter != self.program.instructions().len() {
+            match self.program.instructions()[self.program_counter].instruction() {
+                Instruction::MoveLeft => self.move_head_left()?,
+                Instruction::MoveRight => self.move_head_right()?,
+                Instruction::Increment => self.increment_cell(),
+                Instruction::Decrement => self.decrement_cell(),
+                Instruction::Input => self.read_value(&mut std::io::stdin())?,
+                Instruction::Output => self.print_value(&mut std::io::stdout())?,
+                _ => (),
+            };
+        }
+
+        Ok(())
+    }
+
+    /// Move the head one cell towards the left (start) of the tape
+    fn move_head_left(&mut self) -> Result<(), VMError> {
+        match self.head.checked_sub(1) {
+            Some(new_pos) => {
+                self.head = new_pos;
+                Ok(())
+            }
+            None => {
+                let bad_instruction = self.program.instructions()[self.program_counter].clone();
+                Err(VMError::HeadUnderrun(bad_instruction))
+            }
         }
     }
 
-    /// Print out the intermediate representation of the program
-    pub fn print_program(self, program: &BfProgram) {
-        for instruction in program.instructions().iter() {
-            println!("[{}] {}", program.name().display(), instruction)
+    /// Move the head one cell towards the right (end) of the tape.
+    /// If the head is at the end of the tape and the VM has been instantiated
+    /// with an auto-extending tape, more cells will be added. If not, the VM
+    /// will be sad and will throw an error out.
+    fn move_head_right(&mut self) -> Result<(), VMError> {
+        if self.head == (self.cells.len() - 1) && self.tape_can_grow {
+            self.cells.reserve(1);
+            self.cells.resize(self.cells.capacity(), T::default());
+        }
+
+        self.head += 1;
+        if self.head == self.cells.len() {
+            let bad_instruction = self.program.instructions()[self.program_counter].clone();
+            return Err(VMError::HeadOverrun(bad_instruction));
         }
+        Ok(())
+    }
```

Comment:
You have a doubling of the checks here, perhaps:

```suggestion
    fn move_head_right(&mut self) -> Result<(), VMError> {
        self.head += 1;
        if self.head == self.cells.len() {
            if self.tape_can_grow {
                self.cells.push(T::default());
            } else {
                let bad_instruction = self.program.instructions()[self.program_counter].clone();
                return Err(VMError::HeadOverrun(bad_instruction));
            }
        }
        Ok(())
    }
```

Or similar?